packages/realm/src/app-services/SyncSession.ts

```diff
@@ -441,12 +472,17 @@ export class SyncSession {
    * Can be either:
    *  - `reportIndefinitely` - the registration will stay active until the callback is unregistered
    *  - `forCurrentlyOutstandingWork` - the registration will be active until only the currently transferable bytes are synced
-   * @param callback - Called with the following arguments:
-   * 1. `transferred`: The current number of bytes already transferred
-   * 2. `transferable`: The total number of transferable bytes (the number of bytes already transferred plus the number of bytes pending transfer)
+   * @param callback - see {@link DynamicProgressNotificationCallback} and {@link PartitionBasedSyncProgressNotificationCallback}
+   * @throws if signature of `callback` doesn't match requirements for sync mode
    * @since 1.12.0
    */
   addProgressNotification(direction: ProgressDirection, mode: ProgressMode, callback: ProgressNotificationCallback) {
+    if (this.config.flexible) {
+      assert(
+        isEstimateProgressNotification(callback),
+        `For flexible sync the callback can only take one argument - got ${callback.length} arguments`,
+      );
+    }
```

Comment:
I think this may be a breaking change, since it's still been possible to call the API, and we haven't documented that it should have thrown before. Hmm, could we warn here instead?

Reply:
I agree - this feels more breaking than a deprecation (which might be okay? but should be explicit and released accordingly) if it actually throws instead of simply warning.

Reply:
We could update it to e.g.:
```ts
if (this.config.flexible && !isEstimateProgressNotificationCallback(callback)) {
  /* eslint-disable-next-line no-console */
  console.warn("Deprecated progress notifier used. Please use 'EstimateProgressNotificationCallback' which accepts the estimate as the only argument.");
}
```

But I think we could even skip the warning, since the warning is part of the deprecation in the docs. (We don't usually warn for other deprecations.)

For now I've updated this to just assert that it's a function.